packages/languages/src/browser/language-client-contribution-provider-impl.ts

```diff
@@ -0,0 +1,72 @@
+ * Copyright (C) 2018 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject, postConstruct, named } from 'inversify';
+import { ContributionProvider } from '@theia/core/lib/common';
+import { Disposable } from '@theia/core/lib/common/disposable';
+import { LanguageClientContribution } from './language-client-contribution';
+import { FrontendApplication } from '@theia/core/lib/browser';
+import { LanguageClientContributionProvider } from './language-client-contribution-provider';
+
+@injectable()
+export class LanguageClientContributionProviderImpl implements LanguageClientContributionProvider {
+    @inject(FrontendApplication)
+    protected readonly app: FrontendApplication;
+    @inject(ContributionProvider) @named(LanguageClientContribution)
+    private readonly contributions: ContributionProvider<LanguageClientContribution>;
+
+    protected readonly languageClientContributors = new Map<string, LanguageClientContribution>();
+
+    @postConstruct()
+    collectContributions() {
+        for (const contribution of this.contributions.getContributions()) {
+            this.languageClientContributors.set(contribution.id, contribution);
+        }
+    }
+
+    getLanguageClientContributions(): IterableIterator<LanguageClientContribution> {
+        return this.languageClientContributors.values();
+    }
+
+    registerLanguageClientContribution(clientContribution: LanguageClientContribution): Disposable {
+        const id = clientContribution.id;
+        if (this.isRegistered(id)) {
+            console.warn(`Language contribution with type '${id}' already registered.`);
+            return Disposable.NULL;
+        }
+        this.languageClientContributors.set(clientContribution.id, clientContribution);
+
+        clientContribution.waitForActivation(this.app).then(() =>
+            clientContribution.activate(this.app)
+        );
+
+        return Disposable.create(() => this.unregisterLanguageClientContribution(id));
+    }
+
+    unregisterLanguageClientContribution(id: string): void {
+        this.languageClientContributors.delete(id);
+    }
+
+    private isRegistered(languageId: string): boolean {
+        for (const id of this.languageClientContributors.keys()) {
```

Comment:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/has ?

Reply:
(maybe then we don't need a method for that)